feat: modify rendezvous so that it can be extended

chaitanyaprem opened the pull request:
Modify rendezvous code (using generics) to make it extendable and to support  app specific peerRecord e.g in Waku we want to publish Waku capabilities as part of SignedPeerRecord in rendezvous.

Note that the [SignedEnvelopes](https://github.com/libp2p/specs/blob/377f05a/RFC/0002-signed-envelopes.md) spec allows for CustomPeerRecords to be shared via [Rendezvous](https://github.com/libp2p/specs/blob/master/rendezvous/README.md#the-protocol) . This PR updates rendezvous code to support the same as Waku plans to use rendezvous for capability discovery especially for mix protocol.

Reference [research issue](https://github.com/waku-org/research/issues/118)

Changes

- [x] Update to use Generics instead of PeerRecord
- [x] Update tests
- [x] Add tests to validate by passing a custom PeerRecord 


Thanks a lot @Ivansete-status for all the help in debugging and fixing errors.

Co-authored-by: @Ivansete-status

chaitanyaprem commented on libp2p/protocols/rendezvous/rendezvous.nim:
```diff
@@ -298,30 +303,33 @@ proc advertise*(
 
-method advertise*(
-    rdv: RendezVous, ns: string, ttl: Duration = rdv.minDuration
+method advertise*[T](
```
Thanks, but that did not solve the issue.

Also looks like eventhough they are deprecated, methods with generics seem to work.

richard-ramos reviewed (commented):
If we use any of the operations that accept a generic `[T]`, and we don't set up a peerValidator before that handles this `T`, then peers are not gonna be validated successfully, since we would always be expecting a PeerRecord.

Ideally this type of misconfiguration should not be allowed, which makes me think that the direct usage of these functions should not be encouraged at all.

The same with using the rendezvous codec for custom peer records, since any node using the default codec would expect that a PeerRecord is used and not a custom PeerRecord.

My suggestion would be to extract the generic functionality into it's own `GenericRendezvous` type, and then override methods for your more specialized types, each one with its own custom validator and codec. Here's an example you can run in nim playground  https://play.nim-lang.org/#pasty=JhUDRTEV

```nim

type
  PeerRecordValidator* = proc(spr: seq[byte]): bool

type GenericRendezvous[E] = object
  codec*: string
  peerValidator: PeerRecordValidator

proc new[E](T: typedesc[GenericRendezvous[E]], codec: string, validator: PeerRecordValidator): T =
  GenericRendezvous[E](codec: codec, peerValidator: validator)

proc advertise[T](rdv: GenericRendezvous[T], ns: string, signedPeerRecord: seq[byte]) =
  echo "GenericRendezvous: advertising -- type:", T, " ns:", ns, " bytes: ", signedPeerRecord

proc request[T](rdv: GenericRendezvous[T], ns: string): seq[T] =
  echo "GenericRendezvous: request -- type:", T, " ns:", ns 
  @[]

  
type PeerRecord = object
type CustomType = object
  
type Rendezvous = GenericRendezvous[PeerRecord]
type CustomRendezvous = GenericRendezvous[CustomType]

let validatorRendezvous = proc(spr: seq[byte]): bool =
  return true

let validatorCustomRendezvous = proc(spr: seq[byte]): bool =
  return true

proc advertise(rdv: Rendezvous, ns: string) =
  echo "Calling advertise on Rendezvous"
  procCall GenericRendezvous(rdv).advertise(ns, @[1.byte, 2, 3])
  

let foo = PeerRecord()
let bar = CustomType()

let rdv = Rendezvous.new("/rendezvous/1.0.0", validatorRendezvous)
let customRdv = CustomRendezvous.new("/custom-rendezvous/1.0.0", validatorCustomRendezvous)

rdv.advertise("abc")
```

SionoiS commented:
Nim doesn't have any concept of generic bounds?

richard-ramos commented:
> @SionoiS: Nim doesn't have any concept of generic bounds?

It has concepts, but they're marked as experimental https://nim-lang.org/docs/manual_experimental.html#concepts

chaitanyaprem commented on libp2p/multicodec.nim:
```diff
@@ -397,6 +397,7 @@ const MultiCodecList = [
   ("onion3", 0x01BD),
   ("p2p-circuit", 0x0122),
   ("libp2p-peer-record", 0x0301),
+  ("libp2p-custom-peer-record", 0x0505), #not in multicodec list
```
ah, Thanks did not know there was a private range..makes sense to use that.

chaitanyaprem commented:
> No idea why my vscode nph is screwing up formatting for `tutorial_6` file alone.
> My nph version is `v0.5.1-0-gde5cd48`.

any idea what could be causing the formatting issue in this file? 
are you guys facing the issue as well or is it something i need to fix in my setup? 
For now i have modified the file through vim and fixed the lint issues.
